sympy/polys/matrices/tests/test_nullspace.py

```diff
@@ -0,0 +1,213 @@
+# primitive form. We test two things:
+#
+#   A.nullspace().primitive()[1] == answer.
+#   A.nullspace(divide_last=True) == _divide_last(answer).
+#
+# The nullspace as returned by DomainMatrix and related classes is the
+# transpose of the nullspace as returned by Matrix. Matrix returns a list of
+# of column vectors whereas DomainMatrix returns a matrix whose rows are the
+# nullspace vectors.
+#
+
+
+NULLSPACE_EXAMPLES = [
+
+    (
+        'zz_1',
+         DM([[ 1, 2, 3]], ZZ),
+         DM([[-2, 1, 0],
+             [-3, 0, 1]], ZZ),
+    ),
+
+    (
+        'zz_2',
+         zeros((0, 0), ZZ),
+         zeros((0, 0), ZZ),
+    ),
+
+    (
+        'zz_3',
+        zeros((2, 0), ZZ),
+        zeros((0, 0), ZZ),
+    ),
+
+    (
+        'zz_4',
+        zeros((0, 2), ZZ),
+        eye(2, ZZ),
+    ),
+
+    (
+        'zz_5',
+        zeros((2, 2), ZZ),
+        eye(2, ZZ),
+    ),
+
+    (
+        'zz_6',
+        DM([[1, 2],
+            [3, 4]], ZZ),
+        zeros((0, 2), ZZ),
+    ),
+
+    (
+        'zz_7',
+        DM([[1, 1],
+            [1, 1]], ZZ),
+        DM([[-1, 1]], ZZ),
+    ),
+
+    (
+        'zz_8',
+        DM([[1],
+            [1]], ZZ),
+        zeros((0, 1), ZZ),
+    ),
+
+    (
+        'zz_9',
+        DM([[1, 1]], ZZ),
+        DM([[-1, 1]], ZZ),
+    ),
+
+    (
+        'zz_10',
+        DM([[0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
+            [1, 0, 0, 0, 0, 0, 1, 0, 0, 0],
+            [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
+            [0, 0, 0, 1, 0, 0, 0, 0, 1, 0],
+            [0, 0, 0, 0, 1, 0, 0, 0, 0, 1]], ZZ),
+        DM([[0, 0, -1, 0, 0, 0,  0,  0,  0,  0],
+            [1, 0,  0, 0, 0, 0, -1,  0,  0,  0],
+            [0, 1,  0, 0, 0, 0,  0, -1,  0,  0],
+            [0, 0,  0, 1, 0, 0,  0,  0, -1,  0],
+            [0, 0,  0, 0, 1, 0,  0,  0,  0, -1]], ZZ),
+    ),
+
+]
+
+
+def _to_DM(A, ans):
+    """Convert the answer to DomainMatrix."""
+    if isinstance(A, DomainMatrix):
+        return A.to_dense()
+    elif isinstance(A, DDM):
+        return DomainMatrix(list(A), A.shape, A.domain).to_dense()
+    elif isinstance(A, SDM):
+        return DomainMatrix(dict(A), A.shape, A.domain).to_dense()
+    else:
+        assert False # pragma: no cover
+
+
+def _divide_last(null):
+    """Normalize the nullspace by the rightmost non-zero entry."""
+    null = null.to_field()
+
+    if null.is_zero_matrix:
+        return null
+
+    rows = []
+    for i in range(null.shape[0]):
+        for j in reversed(range(null.shape[1])):
+            if null[i, j]:
+                rows.append(null[i, :] / null[i, j])
+                break
+        else:
+            assert False # pragma: no cover
+
+    return DomainMatrix.vstack(*rows)
+
+
+def _check_primitive(null, null_ans):
+    """Check that the primitive of the answer matches."""
+    null = _to_DM(null, null_ans)
+    cont, null_prim = null.primitive()
+    # Don't check both signs. Fix canonicalization.
```

Comment:
For reference, `primitive` applied to expressions does not extract the sign, only the positive gcd, e.g.
```python
>>> primitive(-2*(x + 1))
(2, -x -1)
```
And `signsimp` will only hold the negative 1 if `evaluate=False`. 

I tend to favor the `canonical` option where, for Relational, we have
```python
>>> (-x < -1).canonical
x > 1
```

Reply:
I think it is reasonable that `primitive` does not extract the sign or in other words only extracts a non-unit gcd.

I'm just not sure what is the appropriate way to make results canonical so that higher level code always sees the same output for each operation. When we have a denominator we can choose to make the denominator have a canonical sign. In general though for a matrix if we have one method return `[1, -1]` and another returns `[-1, 1]` then it is not clear which should be the canonical form when it is permissible to flip the sign.

Perhaps a solution is that since `nullspace` uses `rref_den` it should make the sign canonical with respect to the denominator even though the denominator is going to be discarded. Then the output of nullspace would be canonical after calling primitive.

Reply:
> Perhaps a solution is that since `nullspace` uses `rref_den` it should make the sign canonical with respect to the denominator even though the denominator is going to be discarded. Then the output of nullspace would be canonical after calling primitive.

I've just pushed this change in 84e3e33